controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,1008 @@
+  case FLASH_BASE:
+    ndx = 0;
+    bit = 8;
+    break;
+
+  case GPIO_A_BASE:
+    ndx = 1;
+    bit = 0;
+    break;
+  case GPIO_B_BASE:
+    ndx = 1;
+    bit = 1;
+    break;
+  case GPIO_C_BASE:
+    ndx = 1;
+    bit = 2;
+    break;
+  case GPIO_D_BASE:
+    ndx = 1;
+    bit = 3;
+    break;
+  case GPIO_E_BASE:
+    ndx = 1;
+    bit = 4;
+    break;
+  case GPIO_H_BASE:
+    ndx = 1;
+    bit = 7;
+    break;
+  case ADC_BASE:
+    ndx = 1;
+    bit = 13;
+    break;
+
+  case TIMER2_BASE:
+    ndx = 4;
+    bit = 0;
+    break;
+  case TIMER3_BASE:
+    ndx = 4;
+    bit = 1;
+    break;
+  case TIMER6_BASE:
+    ndx = 4;
+    bit = 4;
+    break;
+  case TIMER7_BASE:
+    ndx = 4;
+    bit = 7;
+    break;
+  case SPI2_BASE:
+    ndx = 4;
+    bit = 14;
+    break;
+  case SPI3_BASE:
+    ndx = 4;
+    bit = 15;
+    break;
+  case UART2_BASE:
+    ndx = 4;
+    bit = 17;
+    break;
+  case UART3_BASE:
+    ndx = 4;
+    bit = 18;
+    break;
+  case UART4_BASE:
+    ndx = 4;
+    bit = 19;
+    break;
+  case I2C1_BASE:
+    ndx = 4;
+    bit = 21;
+    break;
+  case I2C2_BASE:
+    ndx = 4;
+    bit = 22;
+    break;
+  case I2C3_BASE:
+    ndx = 4;
+    bit = 23;
+    break;
+
+  case I2C4_BASE:
+    ndx = 5;
+    bit = 1;
+    break;
+
+  case TIMER1_BASE:
+    ndx = 6;
+    bit = 11;
+    break;
+  case SPI1_BASE:
+    ndx = 6;
+    bit = 12;
+    break;
+  case UART1_BASE:
+    ndx = 6;
+    bit = 13;
+    break;
+  case TIMER15_BASE:
+    ndx = 6;
+    bit = 16;
+    break;
+  case TIMER16_BASE:
+    ndx = 6;
+    bit = 17;
+    break;
+  }
+
+  // If the input address wasn't found then it's definitly
+  // a bug.  I'll just loop forever here causing the code
+  // to crash.  That should make it easier to find the
+  // bug during development.
+  if (ndx < 0) {
+    IntDisable();
+    while (1) {
+    }
+  }
+
+   // Enable the clock of the requested peripherial
+   RCC_Regs *rcc = reinterpret_cast<RCC_Regs *>(RCC_BASE);
+   rcc->periphClkEna[ndx] |= (1<<bit);
+}
+
+/******************************************************************
+ * Interrupt vector table.  The interrupt vector table is a list of
+ * pointers to the various interrupt functions.  It is stored at the
+ * very start of the flash memory.
+ *****************************************************************/
+
+// Fault handlers
+static void fault()
+{
+   while(1){}
+}
+
+static void NMI()           { fault(); }
+static void FaultISR()      { fault(); }
+static void MPUFaultISR()   { fault(); }
+static void BusFaultISR()   { fault(); }
+static void UsageFaultISR() { fault(); }
+static void BadISR()        { fault(); }
+
+extern "C" void Reset_Handler();
+__attribute__((section(".isr_vector"))) void (*const vectors[])() = {
+    // The first entry of the ISR holds the initial value of the
+    // stack pointer.  The ARM processor initializes the stack
+    // pointer based on this address.
+    reinterpret_cast<void (*)()>(&systemStack[SYSTEM_STACK_SIZE]),
```

Comment:
Is it possible to have a "guard page" so we know if we overflow the stack?  (I mean, I know we're bare metal here, so we don't even have pages, but...something like that?)

Reply:
The ARM cortex core has a memory protection unit, but I haven't gotten around to learning how to use it and don't want to start now, at least not as part of this PR.  Merging this HAL is blocking me from other more important stuff, so I really want to get it done.
That said, I think using hardware memory protection around the stack makes a lot of sense and is something I can look into at a later date once some of the basic system function is up and running better.
In the shorter term, I'm planning to add a simple software based stack check that will run periodically from the timer ISR.  Very simple, on init I'll fill the stack with an obvious value, and in the periodic check I'll just make sure that the top N words haven't been touched (for some reasonable value of N).  Far from perfect, but likely to catch most stack overflows.
That's for a future PR though, I don't want to try adding it to this one.